File: MiniDB_Projects/Team_candyman/internal/sql/lexer.go
package sql

import (
	"fmt"
	"strings"
)

type tokenKind int

const (
	tEOF tokenKind = iota
	tIdent
	tNumber
	tString
	tPunct   // ( ) , . ; *
	tOp      // = != <> < <= > >=
	tKeyword // reserved words (upper-cased)
)

type token struct {
	kind tokenKind
	text string
	pos  int
}

var keywords = map[string]bool{
	"CREATE": true, "TABLE": true, "PRIMARY": true, "KEY": true,
	"INSERT": true, "INTO": true, "VALUES": true,
	"DELETE": true, "FROM": true, "WHERE": true,
	"SELECT": true, "JOIN": true, "ON": true, "AND": true, "AS": true,
	"GROUP": true, "BY": true, "EXPLAIN": true,
	"BEGIN": true, "COMMIT": true, "ROLLBACK": true,
	"INT": true, "INTEGER": true, "BIGINT": true,
	"TEXT": true, "STRING": true, "VARCHAR": true,
	"NULL":  true,
	"COUNT": true, "SUM": true, "AVG": true, "MIN": true, "MAX": true,
}

type lexer struct {
	src string
	pos int
}

func newLexer(src string) *lexer { return &lexer{src: src} }

func (l *lexer) errf(format string, a ...any) error {
	return fmt.Errorf("sql: "+format+" (at offset %d)", append(a, l.pos)...)
}

func (l *lexer) tokens() ([]token, error) {
	var toks []token
	for {
		t, err := l.next()
		if err != nil {
			return nil, err
		}
		toks = append(toks, t)
		if t.kind == tEOF {
			return toks, nil
		}
	}
}

func (l *lexer) next() (token, error) {
	for l.pos < len(l.src) {
		c := l.src[l.pos]
		switch {
		case c == ' ' || c == '\t' || c == '\n' || c == '\r':
			l.pos++
			continue
		case c == '-' && l.pos+1 < len(l.src) && l.src[l.pos+1] == '-':
			// line comment
			for l.pos < len(l.src) && l.src[l.pos] != '\n' {
				l.pos++
			}
			continue
		}
		break
	}
	if l.pos >= len(l.src) {
		return token{kind: tEOF, pos: l.pos}, nil
	}
	start := l.pos
	c := l.src[l.pos]
	switch {
	case isIdentStart(c):
		for l.pos < len(l.src) && isIdentPart(l.src[l.pos]) {
			l.pos++
		}
		word := l.src[start:l.pos]
		up := strings.ToUpper(word)
		if keywords[up] {
			return token{kind: tKeyword, text: up, pos: start}, nil
		}
		return token{kind: tIdent, text: word, pos: start}, nil
	case c >= '0' && c <= '9', c == '-' && l.pos+1 < len(l.src) && l.src[l.pos+1] >= '0' && l.src[l.pos+1] <= '9':
		if c == '-' {
			l.pos++
		}
		for l.pos < len(l.src) && l.src[l.pos] >= '0' && l.src[l.pos] <= '9' {
			l.pos++
		}
		return token{kind: tNumber, text: l.src[start:l.pos], pos: start}, nil
	case c == '\'':
		l.pos++ // opening quote
		var sb strings.Builder
		for l.pos < len(l.src) {
			ch := l.src[l.pos]
			if ch == '\'' {
				// '' is an escaped quote
				if l.pos+1 < len(l.src) && l.src[l.pos+1] == '\'' {
					sb.WriteByte('\'')
					l.pos += 2
					continue
				}
				l.pos++
				return token{kind: tString, text: sb.String(), pos: start}, nil
			}
			sb.WriteByte(ch)
			l.pos++
		}
		return token{}, l.errf("unterminated string literal")
	case strings.IndexByte("(),.;*", c) >= 0:
		l.pos++
		return token{kind: tPunct, text: string(c), pos: start}, nil
	case c == '=' || c == '<' || c == '>' || c == '!':
		l.pos++
		if l.pos < len(l.src) {
			two := l.src[start : l.pos+1]
			switch two {
			case "<=", ">=", "<>", "!=":
				l.pos++
				return token{kind: tOp, text: two, pos: start}, nil
			}
		}
		op := l.src[start:l.pos]
		if op == "!" {
			return token{}, l.errf("unexpected '!'")
		}
		return token{kind: tOp, text: op, pos: start}, nil
	default:
		return token{}, l.errf("unexpected character %q", string(c))
	}
}

func isIdentStart(c byte) bool {
	return c == '_' || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}
func isIdentPart(c byte) bool { return isIdentStart(c) || (c >= '0' && c <= '9') }

File: MiniDB_Projects/Team_BlueLeaf/src/parser/lexer.cpp
#include "parser/lexer.h"

#include <cctype>
#include <unordered_map>

#include "common/exception.h"

namespace minidb {

namespace {
// Keyword lookup, keyed by upper-cased text.
const std::unordered_map<std::string, TokenType>& keywords() {
    static const std::unordered_map<std::string, TokenType> kw = {
        {"SELECT", TokenType::SELECT}, {"FROM", TokenType::FROM},
        {"WHERE", TokenType::WHERE},   {"INSERT", TokenType::INSERT},
        {"INTO", TokenType::INTO},     {"VALUES", TokenType::VALUES},
        {"DELETE", TokenType::DELETE}, {"CREATE", TokenType::CREATE},
        {"TABLE", TokenType::TABLE},   {"JOIN", TokenType::JOIN},
        {"ON", TokenType::ON},         {"AND", TokenType::AND},
        {"OR", TokenType::OR},         {"GROUP", TokenType::GROUP},
        {"BY", TokenType::BY},         {"COUNT", TokenType::COUNT},
        {"SUM", TokenType::SUM},       {"AVG", TokenType::AVG},
        {"MIN", TokenType::MIN},       {"MAX", TokenType::MAX},
        {"INT", TokenType::INT_TYPE},  {"INTEGER", TokenType::INT_TYPE},
        {"VARCHAR", TokenType::VARCHAR_TYPE}, {"TEXT", TokenType::VARCHAR_TYPE},
        {"DOUBLE", TokenType::DOUBLE_TYPE},   {"FLOAT", TokenType::DOUBLE_TYPE},
        {"PRIMARY", TokenType::PRIMARY}, {"KEY", TokenType::KEY},
    };
    return kw;
}

std::string to_upper(std::string s) {
    for (char& c : s) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
    return s;
}
} // namespace

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> out;
    std::size_t i = 0, n = sql_.size();

    auto push = [&](TokenType t, std::string text = "") { out.push_back({t, std::move(text)}); };

    while (i < n) {
        char c = sql_[i];

        if (std::isspace(static_cast<unsigned char>(c))) { ++i; continue; }

        // identifiers / keywords
        if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            std::size_t start = i;
            while (i < n && (std::isalnum(static_cast<unsigned char>(sql_[i])) || sql_[i] == '_')) ++i;
            std::string word = sql_.substr(start, i - start);
            auto it = keywords().find(to_upper(word));
            if (it != keywords().end()) push(it->second, word);
            else push(TokenType::IDENTIFIER, word);
            continue;
        }

        // numbers (integer or decimal)
        if (std::isdigit(static_cast<unsigned char>(c)) ||
            (c == '-' && i + 1 < n && std::isdigit(static_cast<unsigned char>(sql_[i + 1])))) {
            std::size_t start = i;
            if (sql_[i] == '-') ++i;
            while (i < n && (std::isdigit(static_cast<unsigned char>(sql_[i])) || sql_[i] == '.')) ++i;
            push(TokenType::NUMBER, sql_.substr(start, i - start));
            continue;
        }

        // single-quoted string literal
        if (c == '\'') {
            ++i;
            std::string s;
            while (i < n && sql_[i] != '\'') s += sql_[i++];
            if (i >= n) throw DBException("Lexer: unterminated string literal");
            ++i;  // closing quote
            push(TokenType::STRING, s);
            continue;
        }

        // operators / punctuation
        switch (c) {
            case '*': push(TokenType::STAR); ++i; break;
            case ',': push(TokenType::COMMA); ++i; break;
            case '(': push(TokenType::LPAREN); ++i; break;
            case ')': push(TokenType::RPAREN); ++i; break;
            case ';': push(TokenType::SEMICOLON); ++i; break;
            case '.': push(TokenType::DOT); ++i; break;
            case '=': push(TokenType::EQ); ++i; break;
            case '<':
                if (i + 1 < n && sql_[i + 1] == '=') { push(TokenType::LTE); i += 2; }
                else if (i + 1 < n && sql_[i + 1] == '>') { push(TokenType::NEQ); i += 2; }
                else { push(TokenType::LT); ++i; }
                break;
            case '>':
                if (i + 1 < n && sql_[i + 1] == '=') { push(TokenType::GTE); i += 2; }
                else { push(TokenType::GT); ++i; }
                break;
            case '!':
                if (i + 1 < n && sql_[i + 1] == '=') { push(TokenType::NEQ); i += 2; }
                else throw DBException("Lexer: unexpected '!'");
                break;
            default:
                throw DBException(std::string("Lexer: unexpected character '") + c + "'");
        }
    }

    push(TokenType::END);
    return out;
}

} // namespace minidb
